CSharpMath.Editor/MathKeyboard.cs

```diff
@@ -56,20 +56,27 @@ public class MathKeyboard<TFont, TGlyph> where TFont : IFont<TGlyph> {
       }
     }
     MathKeyboardCaretState _caretState;
+    /// <summary>Adjust a placeholder and invoke a display/redraw when CaretState or display is initialized or changed.</summary>
+    private void ProcessCaretState() {
+      if (MathList.AtomAt(_insertionIndex) is Atoms.Placeholder placeholder) {
+        (placeholder.Nucleus, placeholder.Color) =
+          (_caretState == MathKeyboardCaretState.Shown)
+          ? (LaTeXSettings.PlaceholderActiveNucleus, LaTeXSettings.PlaceholderActiveColor)
+          : (LaTeXSettings.PlaceholderRestingNucleus, LaTeXSettings.PlaceholderRestingColor);
+      }
+      RecreateDisplayFromMathList();
+      RedrawRequested?.Invoke(this, EventArgs.Empty);
+    }
+
     public MathKeyboardCaretState CaretState {
       get => _caretState;
       set {
         blinkTimer.Stop();
         blinkTimer.Start();
-        if (value != MathKeyboardCaretState.Hidden &&
-           MathList.AtomAt(_insertionIndex) is Atoms.Placeholder placeholder) 
-          (placeholder.Nucleus, placeholder.Color, _caretState) =
-            value == MathKeyboardCaretState.TemporarilyHidden
-            ? (LaTeXSettings.PlaceholderRestingNucleus, LaTeXSettings.PlaceholderRestingColor, MathKeyboardCaretState.TemporarilyHidden)
-            : (LaTeXSettings.PlaceholderActiveNucleus, LaTeXSettings.PlaceholderActiveColor, MathKeyboardCaretState.ShownThroughPlaceholder);
-        else _caretState = value;
-        RecreateDisplayFromMathList();
-        RedrawRequested?.Invoke(this, EventArgs.Empty);
+        if (value != _caretState) {
+          _caretState = value;
+          ProcessCaretState();
```

Comment:
move it out?

Reply:
I mean only `ProcessCaretState();` For some reason comment highlighted 3 lines instead of 1 as I selected.

Reply:
Can you clarify? Move it out of the `if`? The purpose is avoid the heavy stuff inside ProcessCaretState when CaretState is unchanged.

Reply:
yep, but latest commit adds call to `ProcessCaretState();` almost everywhere anyway

Reply:
Hmm maybe I was wrong about this change then. I had the impression (from testing the version deployed to nuget) that the timer would often set the CaretState to itself and cause a redraw. But that doesn't seem to be the case at least in the current state of this branch.